tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs

```diff
@@ -0,0 +1,147 @@
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Authentication;
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Specifications.auth
+{
+    public class AuthTestRunner
+    {
+        [SkippableTheory]
+        [ClassData(typeof(TestCaseFactory))]
+        public void RunTestDefinition(JsonDrivenTestCase testCase)
+        {
+            var definition = testCase.Test;
+            JsonDrivenHelper.EnsureAllFieldsAreValid(definition, "description", "uri", "valid", "credential");
+
+            MongoCredential mongoCredential = null;
+            Exception parseException = null;
+            try
+            {
+                var connectionString = (string)definition["uri"];
+                mongoCredential = MongoClientSettings.FromConnectionString(connectionString).Credential;
+            }
+            catch (Exception ex)
+            {
+                parseException = ex;
+            }
+
+            if (parseException == null)
+            {
+                AssertValid(mongoCredential, definition);
+            }
+            else
+            {
+                AssertInvalid(parseException, definition);
+            }
+        }
+
+        private void AssertValid(MongoCredential mongoCredential, BsonDocument definition)
+        {
+            if (!definition["valid"].ToBoolean())
+            {
+                throw new AssertionException($"The connection string '{definition["uri"]}' should be invalid.");
+            }
+
+            var expectedCredential = definition["credential"] as BsonDocument;
+            if (expectedCredential != null)
+            {
+                JsonDrivenHelper.EnsureAllFieldsAreValid(expectedCredential, "username", "password", "source", "mechanism", "mechanism_properties");
+                mongoCredential.Username.Should().Be(ValueToString(expectedCredential["username"]));
+#pragma warning disable 618
+                mongoCredential.Password.Should().Be(ValueToString(expectedCredential["password"]));
+#pragma warning restore 618
+                mongoCredential.Source.Should().Be(ValueToString(expectedCredential["source"]));
+                mongoCredential.Mechanism.Should().Be(ValueToString(expectedCredential["mechanism"]));
+
+                var authenticator = mongoCredential.ToAuthenticator();
+                if (expectedCredential.TryGetValue("mechanism_properties", out var expectedMechanismProperties))
+                {
+                    if (expectedMechanismProperties.IsBsonNull)
```

Comment:
Since I am proposing that the reflector reflect of `GssapiAuthenticator` instead of an interface, just refactor lines 79-112 to:

    if (authenticator is GssapiAuthenticator gssapiAuthenticator)
    {
        if (expectedMechanismProperties.IsBsonNull)
        {
            var serviceName = gssapiAuthenticator._mechanism_serviceName();
            serviceName.Should().Be("mongodb"); // The default is "mongodb".
            var canonicalizeHostName = gssapiAuthenticator._mechanism_canonicalizeHostName();
            canonicalizeHostName.Should().BeFalse(); // The default is "false".
        }
        else
        {
            foreach (var expectedMechanismProperty in expectedMechanismProperties.AsBsonDocument)
            {
                var mechanismName = expectedMechanismProperty.Name;
                switch (mechanismName)
                {
                    case "SERVICE_NAME":
                        var serviceName = gssapiAuthenticator._mechanism_serviceName();
                        serviceName.Should().Be(ValueToString(expectedMechanismProperty.Value));
                        break;
                    case "CANONICALIZE_HOST_NAME":
                        var canonicalizeHostName = gssapiAuthenticator._mechanism_canonicalizeHostName();
                        canonicalizeHostName.Should().Be(expectedMechanismProperty.Value.ToBoolean());
                        break;
                    default:
                        throw new Exception($"Invalid mechanism property '{mechanismName}'.");
                }
            }
        }
    }
    else
    {
        // Other authenticators do not contain mechanism properties
    }

Reply:
Done